### PR TITLE
Add group mgt v2 audit logger

### PR DESCRIPTION
**Pull request by VivekVinushanth**
### Proposed changes in this pull request
This PR onboard v2 audit logger for group management operations 

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/listeners/GroupManagementV2AuditLogger.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            return isEnableV2AuditLogs();
+        }
+        return false;
+    }
+
+    @Override
+    public boolean postAddGroup(String groupName, String groupId, List<String> userIds, List<Claim> claims,
+                                UserStoreManager userStoreManager) {
+
+        if (!isEnable()) {
```
Comment (piraveena) [inline components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/listeners/GroupManagementV2AuditLogger.java:67]: do we need this enable check?

File: components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/listeners/GroupManagementV2AuditLogger.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        triggerAuditLogEvent(auditLogBuilder);
+        return true;
+    }
+
+    @Override
+    public boolean postUpdateUserListOfGroup(String groupId, List<String> deletedUserIds, List<String> newUserIds,
+                                             UserStoreManager userStoreManager) {
+
+        if (!isEnable()) {
+            return true;
```
Comment (piraveena) [inline components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/listeners/GroupManagementV2AuditLogger.java:114]: check the previous comment

File: components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/listeners/GroupManagementV2AuditLogger.java (hunk 1 above)
Comment (VivekVinushanth) [reply components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/listeners/GroupManagementV2AuditLogger.java:68]: At this point we need to.
1. At kernel, we use **AbstractGroupOperationEventListener** which does not have the isEnabled check
2. **AbstractIdentityGroupOperationEventListener** which the logger extends from can not be used at kernel as it is from framework and hence introduce cyclic dependency.
3. This audit logger can be independently disabled. But audit loggers are enabled by default and we rely on the config to enable. We check if the AbstractIdentityGroupOperationEventListener is enabled and then proceed to check if **V2auditlogging** is enabled as well. We do this check at each method, to ensure unless this V2 is enabled, we should not log.

File: components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/listeners/GroupManagementV2AuditLogger.java (hunk 2 above)
Comment (VivekVinushanth) [reply components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/listeners/GroupManagementV2AuditLogger.java:115]: https://github.com/wso2/carbon-identity-framework/pull/5593#discussion_r1538620086

File: components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/internal/UserMgtDSComponent.java [hunk 3]
```diff
@@ -135,6 +136,16 @@ protected void activate(ComponentContext ctxt) {
                 }
             }
 
+            serviceRegistration = ctxt.getBundleContext().registerService(GroupOperationEventListener.class.getName(),
+                    new GroupManagementV2AuditLogger(), null);
+            if (serviceRegistration == null) {
+                log.error("Error while registering GroupManagementV2AuditLogger.");
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("GroupManagementV2AuditLogger successfully registered.");
```
Comment (piraveena) [inline components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/internal/UserMgtDSComponent.java:145]: no need to check for isDebugEnabled

**Review by piraveena — APPROVED** [review]

**Pull request merged**